### DISCUSSION TITLE
I2S Webserver 2 channels 44.1KHz, possible?

### DISCUSSION THREAD

**Discussion by DynamoBen** [pschatzmann/arduino-audio-tools#1445] [Q&A] [answered]
I've been experimenting with the I2S webserver_wav at 44.1KHz stereo with an ESP32 WROOM. The sound quality is poor and inconsistent (some static and stream starts and stops). I'm experimented some with buffer sizes after reading the wiki on Performance but didn't get much improvement. Am I exceeding the abilities of the ESP32 in this case?

**Comment by pschatzmann** [comment, marked as answer]
https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working

I guess you just forgot to set the log level to warning. I was just running the [test sketch with a sine generator](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-webserver/streams-generator-webserver_wav/streams-generator-webserver_wav.ino) at 44100 stereo and I can not reproduce your issues. Even 48000 works just fine...

**Comment by DynamoBen** [comment]
I didn't forget to set the log level to warning; I've tried both warning and error...no change. I've tried modifying URLStreamBuffer and I2S buffer settings and counts...no change.

The sine example starts and stops at 44.1KHz and 2 channels. I2S skips, starts and stops. I've verified that the I2S signal from the device is not experiencing these issues. Other things worth trying?

Also when using I2S, the rx pin must be set manually in the config even when using the default ESP32 pins.

**Comment by pschatzmann** [comment]
Did you try the test sketch with the sine generator ? I was using Chrome to test...

**Reply by DynamoBen** [reply]
Yes, and at 44.1KHz and 2 channels it starts and stops repeatedly every few seconds. I'm using one of [these ](https://www.adafruit.com/product/5400)if that helps.

**Reply by pschatzmann** [reply]
Strange: since the above test sketch (with the sine generator) is working for me and not for you then you would need to find out where the difference is

- AudioTools Software (Version) - try to use the latest updated version
- Browser did you test with Chrome as well ?
- ESP32 did you try with another ESP32 ?

Did  you set the Log Level on the Test Sketch to Warning as well ? I can only reproduce your issue when the log level is Info...

ps. try to move the ESP32 close to the Router and make sure that you have a good WiFi connection. And I assume that you did not change the default CPU freq of 240MHz...

**Reply by DynamoBen** [reply]
I pulled latest on AudioTools, although the version I had been using was only a week old. I set the test sketch log level to warning as well. CPU freq has not changed. I haven't tried another ESP32 because I don't have any spares. I'm testing with Chrome and VLC. I've also downgraded the arduino-esp32 board version from 3.0 alpha, to 2x. 

The ESP32 is just 10' from the router and has line of sight. I've also tested with a second router made by another manufacturer which was just 2' away. 

I was able to get up to 48KHz with I2S, but skips and halts persist. I'd like to try some bigger buffer sizes and counts, can you make some recommendations? The dev board I'm using has 2 MB PSRAM .

**Reply by pschatzmann** [reply]
The fact that the audio is breaking up indicates that you are not able to provide the audio fast enough, so I am afraid in this case buffering will not help and you will need to reduce the sample rate or channels to reduce the amount of transmitted data. 

If you can output 48000 via I2s w/o any problems the bottleneck seems to be the network speed. So you could also try to send encoded/compressed audio which also reduces the amount of data.

I am still confused that I can't reproduce your issue: can you try to add WiFi.setSleep(false); to your sketch ?

**Reply by pschatzmann** [reply]
I have added the new method setCopyBufferSize() to the AudioServer class which lets you define the amount of data that is copied in each loop step.

**Reply by DynamoBen** [reply]
I've gotten new hardware, I'm now using a ESP32-S3-WROOM-1-N8R8 Module with 8MB PSRAM and 8MB Flash. I'm connecting to an AP where the ESP32 is the only wireless device. I've also switched to MP3 encoding in an attempt to reduce any network speed bottlenecks. 

With audio set to 1 channel at 44.1KHz it works well and is mostly stable. However increasing to 2 channels the audio breaks up (stumbles, starts/stops) and sounds like its sped up, I'm assuming this is an indication of missing frames of data?

I've tried `setCopyBufferSize()` and didn't notice a difference (see how I'm using it below). Manually setting `WiFi.setSleep(false)` doesn't seem to have an effect, which is expected since that is already set in the code. 

Not sure where or how to confirm the bottleneck. If it's WIFI I'd need to investigate ways to improve the ESP32s Wi-Fi performance. If it's code execution I guess I'd need to look at I2S vs MP3 encoding vs Webserver and see where things are getting lost along the way.

```
  // Start audio webserver
  server.begin(i2sStream, config); 
  server.setCopyBufferSize(4096);
```

**Reply by pschatzmann** [reply]
Did you test with the sine generator or the input of data from I2S ? The difference is that the sine generator provides the samples at max speed and the I2S only at the speed of the defined sample rate.

What exactly is your input device ?

**Reply by DynamoBen** [reply]
I've been focused on I2S mainly, the source is an si4735. I've checked the I2S output with another I2S device on the buss and the audio is good. 

I just ran the sine generator with 2 channels at 44.1KHz and logging at Warning. After ~30-40 seconds the stream has a brief stops every ~10 seconds in Chrome (in VLC the stops are further apart). 

Another interesting discovery is that the wav webserver with I2S as the source 2 channels and 44.1KHz runs consistently and sounds great.
